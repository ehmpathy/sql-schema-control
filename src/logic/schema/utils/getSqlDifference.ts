import { diffStringsUnified } from 'jest-diff';

export const getSqlDifference = ({
  oldSql,
  newSql,
}: {
  oldSql: string;
  newSql: string;
}) => {
  // check strings are not equivalent
  if (oldSql === newSql) return null; // i.e., no difference

  // compute the differences
  const sqlDiffString = diffStringsUnified(oldSql.trim(), newSql.trim());

  // return the diff string
  return sqlDiffString;
};
