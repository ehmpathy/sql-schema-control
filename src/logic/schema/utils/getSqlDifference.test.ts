import { getSqlDifference } from './getSqlDifference';

describe('getSqlDifference', () => {
  it('should find that a colored diff of changes if sql is different', () => {
    const oldSql = `
CREATE TABLE IF NOT EXISTS some_table (
id              INT(11) PRIMARY KEY AUTO_INCREMENT,
created_at      DATETIME DEFAULT CURRENT_TIMESTAMP,
) ENGINE = InnoDB;
    `;
    const newSql = `
CREATE TABLE IF NOT EXISTS some_table (
id              BIGINT PRIMARY KEY AUTO_INCREMENT,
created_at      DATETIME(3) DEFAULT CURRENT_TIMESTAMP
) ENGINE = InnoDB;
    `;

    // get the diff
    const result = getSqlDifference({ oldSql, newSql });
    expect(typeof result).toEqual('string');
    expect(result).toMatchSnapshot(); // result is purely visual, so log an example of it
  });
  it('should return null if there is no difference', () => {
    const sql = `
CREATE TABLE IF NOT EXISTS some_table (
  id              INT(11) PRIMARY KEY AUTO_INCREMENT,
  created_at      DATETIME DEFAULT CURRENT_TIMESTAMP,
) ENGINE = InnoDB;
    `;

    // get the diff
    const result = getSqlDifference({ oldSql: sql, newSql: sql });
    expect(result).toEqual(null);
  });
});
