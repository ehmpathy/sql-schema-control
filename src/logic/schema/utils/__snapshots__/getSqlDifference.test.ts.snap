// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getSqlDifference should find that a colored diff of changes if sql is different 1`] = `
"[32m- Expected[39m
[31m+ Received[39m

[2m  CREATE TABLE IF NOT EXISTS some_table ([22m
[32m-   id              INT[7m(11)[27m PRIMARY KEY AUTO_INCREMENT,[39m
[31m+   id              [7mBIG[27mINT PRIMARY KEY AUTO_INCREMENT,[39m
[32m-   created_at      DATETIME DEFAULT CURRENT_TIMESTAMP[7m,[27m[39m
[31m+   created_at      DATETIME[7m(3)[27m DEFAULT CURRENT_TIMESTAMP[39m
[2m  ) ENGINE = InnoDB;[22m"
`;
